cpp/src/parquet/encoding.h

```diff
@@ -168,6 +300,40 @@ class TypedDecoder : virtual public Decoder {
     }
     return num_values;
   }
+
+  /// \brief Decode into an ArrayBuilder or other accumulator
+  ///
+  /// \return number of values decoded
+  virtual int DecodeArrow(int num_values, int null_count, const uint8_t* valid_bits,
+                          int64_t valid_bits_offset,
+                          typename EncodingTraits<DType>::Accumulator* out) = 0;
+
+  /// \brief Decode into an ArrayBuilder or other accumulator ignoring nulls
+  ///
+  /// \return number of values decoded
+  virtual int DecodeArrowNonNull(int num_values,
+                                 typename EncodingTraits<DType>::Accumulator* out) {
+    const uint8_t valid_bits = 0;
+    return DecodeArrow(num_values, 0, &valid_bits, 0, out);
+  }
+
+  /// \brief Decode into a DictionaryBuilder
+  ///
+  /// \return number of values decoded
+  virtual int DecodeArrow(int num_values, int null_count, const uint8_t* valid_bits,
+                          int64_t valid_bits_offset,
+                          typename EncodingTraits<DType>::DictAccumulator* builder) {
+    ParquetException::NYI("FIXME");
+  }
```

Comment:
?

Reply:
sorry, this is supposed to be pure virtual. This exception was a placeholder when this method was not implemented for *all* types.